### ISSUE TITLE
OpenSSL 3.x and EVP_CIPHER_CTX_FLAG_WRAP_ALLOW

### ISSUE THREAD

**Issue by ifranzki** [openssl/openssl#22404]
I noticed a slight behavior change between OpenSSL 1.1.1 and OpenSSL 3.x. 

In OpenSSL 1.1.1, when using `EVP_CipherInit_ex()` with `EVP_aes_256_wrap_pad()` as cipher, and the context does not have flag `EVP_CIPHER_CTX_FLAG_WRAP_ALLOW` set, it fails with `EVP_R_WRAP_MODE_NOT_ALLOWED`.

In OpenSSL 3.x it does not seem to care if the flag is set or not, at least for provided PKEYs. I still find the code that checks that flag in `evp_cipher_init_internal()`, but only for legacy PKEYs.

Is this intended? 

Certainly, it does not hurt if the flag is set in OpenSSL 3.0, so therefore it is backward compatible. 

I noticed this difference, because I tried to make a program that works on OpenSSL 3.0 to also work on OpenSSL 1.1.1, and it failed because I did not set that flag. Not a real problem, just wondering if that flag was used for a good reason, and now it is missing the check ?

**Comment by t8m** [comment]
Yeah, we now do not require the flag to be set anymore. It was more like _I know what I am doing_ kind of flag so ignoring it and not requiring it anymore does not seem to be a problem to me. Perhaps some documentation could be improved?

**Comment by paulidale** [comment]
Agreed, it's a documentation problem.

**Comment by ifranzki** [comment]
The only 2 places where flag EVP_CIPHER_CTX_FLAG_WRAP_ALLOW  is mentioned in the docs is 
https://www.openssl.org/docs/man3.1/man3/EVP_CIPHER_CTX_set_flags.html 
 
_FLAGS
...
EVP_CIPHER_CTX_FLAG_WRAP_ALLOW
Used for Legacy purposes only. This flag needed to be set to indicate the cipher handled wrapping._

and

_NOTES
...
There are some differences between functions EVP_CipherInit() and EVP_CipherInit_ex(), significant in some circumstances. EVP_CipherInit() fills the passed context object with zeros. As a consequence, EVP_CipherInit() does not allow step-by-step initialization of the ctx when the key and iv are passed in separate calls. It also means that the flags set for the CTX are removed, and it is especially important for the EVP_CIPHER_CTX_FLAG_WRAP_ALLOW flag treated specially in EVP_CipherInit_ex()._

The first one mentions _Legacy purposes_ which could be understood as '_only used for legacy PKEYs_'. But I guess we could clarify this a bit more.

The second one mentions it in regards of `EVP_CipherInit_ex()` vs `EVP_CipherInit()`. And yes, the code still treats this flag special in `EVP_CipherInit_ex()` and retains that flag when a new cipher is set for the context, but also only for legacy PKEYs. For provided PKEYs it does reset the flags, but also does not care about that flag later on.
